// @imask/svelte/dist/svelte-imask.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('imask')) :
  typeof define === 'function' && define.amd ? define(['exports', 'imask'], factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory(global.SvelteIMask = {}, global.IMask));
})(this, (function (exports, IMask) { 'use strict';

  function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

  var IMask__default = /*#__PURE__*/_interopDefault(IMask);

  function fireEvent(el, eventName, data) {
    const e = document.createEvent('CustomEvent');
    e.initCustomEvent(eventName, true, true, data);
    el.dispatchEvent(e);
  }
  function initMask(el, opts) {
    const maskRef = opts instanceof IMask__default.default.InputMask ? opts : IMask__default.default(el, opts);
    return maskRef.on('accept', () => fireEvent(el, 'accept', maskRef)).on('complete', () => fireEvent(el, 'complete', maskRef));
  }
  function IMaskAction(el, opts) {
    let maskRef = opts && initMask(el, opts);
    function destroy() {
      if (maskRef) {
        maskRef.destroy();
        maskRef = undefined;
      }
    }
    function update(opts) {
      if (!opts) return destroy();
      if (opts instanceof IMask__default.default.InputMask) maskRef = opts;else if (maskRef) maskRef.updateOptions(opts);else maskRef = initMask(el, opts);
    }
    return {
      update,
      destroy
    };
  }

  Object.defineProperty(exports, 'IMask', {
    enumerable: true,
    get: function () { return IMask__default.default; }
  });
  exports.imask = IMaskAction;

}));
